pysmt/simplifier.py

```diff
@@ -280,47 +280,37 @@ def walk_exists(self, formula, args, **kwargs):
         return self.manager.Exists(varset, sf)
 
     def walk_plus(self, formula, args, **kwargs):
-        is_real = any(x.is_real_constant() for x in args)
-        is_int = any(x.is_int_constant() for x in args)
-        assert not (is_real and is_int)
-
-        if all(x.is_constant() for x in args):
-            res = sum(x.constant_value() for x in args)
-            if is_real:
-                return self.manager.Real(res)
+        new_args = []
+        constant_add = 0
+        stack = list(args)
+        ttype = self.env.stc.get_type(args[0])
+        is_algebraic = False
+        while len(stack) > 0:
+            x = stack.pop()
+            if x.is_constant():
+                if x.is_algebraic_constant():
+                    is_algebraic = True
+                constant_add += x.constant_value()
```

Comment:
My key question is whether Z3 is doing the right thing here, and overloading the + . A test would be the easiest way to show this.

Reply:
I have added a simple test case for both addition and multiplication in test/test_simplify.py" `test_plus_algebraic` and `test_times_algebraic`.    
I have annotated them with `@skipIfSolverNotAvailable("z3")`, since if I am not mistaken algebraic constants are instantiated only when z3 is available.

Reply:
Yes, that is correct. Algebraic constants are currently defined by using the internal z3 representation. We have no other way to represent them, so z3 is a requirement.